### PR TITLE
Change description ILMBASE_FORCE_CXX03

### PR DESCRIPTION
**Pull request by Jamaika1**

### CODE REVIEW

File: IlmBase/IlmThread/IlmThread.h [hunk 1]
```diff
@@ -94,19 +94,25 @@
...
 #       include <windows.h>
 #       include <process.h>
-#   elif HAVE_PTHREAD
-#      include <pthread.h>
+#   else
+#      include <pthread.h>
+#   endif
+#   endif
+#else
+#   if defined(__MINGW32__) || defined(__MINGW64__)
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThread.h]: I think this is the part of the diff that fixes compatibility with mingw?

File: IlmBase/IlmThread/IlmThreadPool.cpp [hunk 2]
```diff
@@ -42,40 +42,44 @@
...
-#ifndef ILMBASE_FORCE_CXX03
+#if !ILMBASE_FORCE_CXX03
 # include <memory>
-# include <atomic>
-# include <thread>
-#else
-# ifndef _WIN32
-#  include <unistd.h>
+# include <atomic>
+# if defined(__MINGW32__) || defined(__MINGW64__)
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThreadPool.cpp:50]: also this

File: IlmBase/IlmThread/IlmThreadPool.cpp [hunk 3]
```diff
@@ -485,7 +489,7 @@ TaskGroup::Data::~Data ()
 
     isEmpty.wait ();
 
-#ifdef ENABLE_SEM_DTOR_WORKAROUND
+#if ENABLE_SEM_DTOR_WORKAROUND
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThreadPool.cpp:492]: Is ENABLE_SEM_DTOR_WORKAROUND guaranteed to be always defined? The existing code assumes that it is not defined at all if the workaround is not desired. There is no guarantee that #if will result in a value of zero if it is not defined; this behavior varies between preprocessor implementations.

File: IlmBase/IlmThread/IlmThreadPool.cpp [hunk 4]
```diff
@@ -504,14 +508,14 @@ TaskGroup::Data::~Data ()
...
-TaskGroup::Data::addTask () 
+TaskGroup::Data::addTask ()
 {
     //
     // in c++11, we use an atomic to protect numPending to avoid the
     // extra lock but for c++98, to add the ability for custom thread
     // pool we add the lock here
     //
-#ifdef ILMBASE_FORCE_CXX03
+#if ILMBASE_FORCE_CXX03
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThreadPool.cpp:518]: We didn't intend for this to be always defined, which #if requires. Is there a reason why you might wish to force it to be zero or one?

File: IlmBase/IlmThread/IlmThreadPool.cpp [hunk 5]
```diff
@@ -536,34 +540,34 @@ TaskGroup::Data::removeTask ()
 
     // Further update:
     //
-    // we could remove this if it is a new enough glibc, however 
+    // we could remove this if it is a new enough glibc, however
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThreadPool.cpp:543]: If there's a desire to clean up white space we always appreciate that such cleaning be part of a separate PR that only does cleaning. It's difficult to be sure a review is thorough when there are many of these to be checked.

File: IlmBase/IlmThread/IlmThread.cpp [hunk 6]
```diff
@@ -43,11 +43,11 @@
 
 #include "IlmBaseConfig.h"
 #include "IlmThread.h"
-#include "Iex.h"
+#include "../Iex/Iex.h"
```
Comment (meshula) [inline IlmBase/IlmThread/IlmThread.cpp:46]: Did you encounter a bug in include path set up somewhere? We aren't intending that such relative paths should be necessary.

**Review by meshula — CHANGES_REQUESTED** [review]
Hi, thanks for the PR. I have worked through part of it, and I think I've spotted what the intent of the change is. I'm requesting that the comments I've listed here be addressed and the PR amended in order to continue. Basically, the high volume of white space changes makes it difficult to be sure all code changes have been reviewed. Thanks for your consideration.

File: IlmBase/IlmThread/IlmThread.h (hunk 1 above)
Comment (Jamaika1) [reply IlmBase/IlmThread/IlmThread.h]: I am asking for your understanding. I'm just an amateur.
In my opinion, MINGW makes sense here because the definition of std :: thread is used below. Looks worse adding MINGW for std :: mutex under windows for gcc as below is cut. Struct semaphore {} works badly.

PS. New definition std::jthread has appeared. Apparently more modern than std::thread. For c++17 I managed to compile. However for c++11 I couldn't find the jthread.h plugin.

File: IlmBase/IlmThread/IlmThreadPool.cpp (hunk 3 above)
Comment (Jamaika1) [reply IlmBase/IlmThread/IlmThreadPool.cpp:492]: I added one to the ENABLE_SEM_DTOR_WORKAROUND definition to make it even active in line 66.

File: IlmBase/IlmThread/IlmThread.cpp (hunk 6 above)
Comment (Jamaika1) [reply IlmBase/IlmThread/IlmThread.cpp:46]: I was more comfortable with GCC.

File: IlmBase/IlmThread/IlmThreadPool.cpp (hunk 5 above)
Comment (Jamaika1) [reply IlmBase/IlmThread/IlmThreadPool.cpp:543]: I only used the Codeblocs editor which did it myself.

File: IlmBase/IlmThread/IlmThread.h (hunk 1 above)
Comment (meshula) [reply IlmBase/IlmThread/IlmThread.h]: jthread is in c++20, which is beyond our target. This change makes sense, my question was if the main topic of your PR is to select the threading api with more detail than the existing code.

File: IlmBase/IlmThread/IlmThread.cpp (hunk 6 above)
Comment (meshula) [reply IlmBase/IlmThread/IlmThread.cpp:46]: We build with gcc as well, but didn't encounter this problem before. Perhaps you could file a new issue, with your build log in it, because the issue of finding Iex is different than solving the mingw thread issues.

File: IlmBase/IlmThread/IlmThreadPool.cpp (hunk 5 above)
Comment (meshula) [reply IlmBase/IlmThread/IlmThreadPool.cpp:543]: I noticed there are a lot of questions about this in the CodeBlocks forums, maybe there's an answer there about how to turn off that feature. It's nice to have that feature when you are working in your codebase, but when working in open source it makes things more confusing for your collaborators.

File: IlmBase/IlmThread/IlmThread.cpp (hunk 6 above)
Comment (Jamaika1) [reply IlmBase/IlmThread/IlmThread.cpp:46]: Ok thanks for the answers. I will just add that there is a bug in IlmThreadMutex.cpp. I changed the record to
```
# if defined _WIN32 || defined _WIN64
```
when ILMBASE_FORCE_CXX03 is active.
